test/e2e/network/service.go

```diff
@@ -1623,6 +1623,126 @@ var _ = SIGDescribe("ESIPP [Slow] [DisabledForLargeClusters]", func() {
+			Expect(err).NotTo(HaveOccurred())
+		}()
+		execPod, err := cs.CoreV1().Pods(ns).Get(execPodName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(framework.AffinityTestFromPod(execPod, nodeIpPort)).To(BeFalse(), "Affinity shouldn't hold.")
+		svc = jig.UpdateServiceOrFail(svc.Namespace, svc.Name, func(svc *v1.Service) {
+			svc.Spec.SessionAffinity = v1.ServiceAffinityClientIP
+		})
+		Expect(framework.AffinityTestFromPod(execPod, nodeIpPort)).To(BeTrue(), "Affinity didn't hold.")
+	})
+
+	It("should be able to enable session affinity for LoadBalancer service", func() {
+		ns := f.Namespace.Name
+		numPods, servicePort := 3, 80
+		serviceName := "service"
+
+		By("creating service in namespace " + ns)
+		_, _, err := framework.StartServeHostnameService(cs, f.InternalClientset, ns, serviceName, servicePort, numPods)
+		defer func() {
+			framework.StopServeHostnameService(cs, f.InternalClientset, ns, serviceName)
+		}()
+		jig := framework.NewServiceTestJig(cs, serviceName)
+		svc := jig.UpdateServiceOrFail(ns, serviceName, func(svc *v1.Service) {
+			svc.Spec.Type = v1.ServiceTypeLoadBalancer
+			svc.Spec.SessionAffinity = v1.ServiceAffinityNone
+			svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeCluster
+		})
+		By("waiting for loadbalancer for service " + ns + "/" + serviceName)
+		svc = jig.WaitForLoadBalancerOrFail(ns, serviceName, loadBalancerCreateTimeout)
+		jig.SanityCheckService(svc, v1.ServiceTypeLoadBalancer)
+		serviceLBNames = append(serviceLBNames, cloudprovider.GetLoadBalancerName(svc))
+		defer func() {
+			jig.ChangeServiceType(svc.Namespace, svc.Name, v1.ServiceTypeClusterIP, loadBalancerCreateTimeout)
+			Expect(cs.CoreV1().Services(svc.Namespace).Delete(svc.Name, nil)).NotTo(HaveOccurred())
+		}()
+		ingressIP := framework.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
+		port := strconv.Itoa(int(svc.Spec.Ports[0].Port))
+		lbIpPort := net.JoinHostPort(ingressIP, port)
+
+		execPodName := framework.CreateExecPodOrFail(cs, ns, "execpod-affinity", nil)
+		defer func() {
+			framework.Logf("Cleaning up the exec pod")
+			err := cs.CoreV1().Pods(ns).Delete(execPodName, nil)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+		execPod, err := cs.CoreV1().Pods(ns).Get(execPodName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(framework.AffinityTestFromPod(execPod, lbIpPort)).To(BeFalse(), "Affinity shouldn't hold.")
+
+		svc = jig.UpdateServiceOrFail(svc.Namespace, svc.Name, func(svc *v1.Service) {
+			svc.Spec.SessionAffinity = v1.ServiceAffinityClientIP
+		})
+		Expect(framework.AffinityTestFromPod(execPod, lbIpPort)).To(BeTrue(), "Affinity didn't hold.")
+		svc = jig.UpdateServiceOrFail(svc.Namespace, svc.Name, func(svc *v1.Service) {
+			svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeLocal
+		})
+		Expect(framework.AffinityTestFromPod(execPod, lbIpPort)).To(BeTrue(), "Affinity didn't hold.")
```

Comment:
It takes sometime for the ExternalTrafficPolicy change to be propagated across nodes, which means this step maybe run before the propagation finishes.

Reply:
LoadBalancer tests split into two which should fix the propagation problem.